packages/form-js-playground/test/spec/JSONEditor.spec.js

```diff
@@ -0,0 +1,187 @@
+            info: 'Written in Service Task',
+            detail: 'Process_1'
+          },
+          {
+            name: 'Variable2',
+            info: 'Written in Service Task',
+            detail: 'Process_1'
+          }
+        ]);
+      }).not.to.throw();
+    });
+
+
+    it('should suggest updated variables', async function() {
+
+      // given
+      const value = '{ "foo": "bar" }';
+
+      const editor = new JSONEditor();
+
+      editor.setValue(value);
+
+      const cm = getCm(editor);
+
+      // move cursor to the end
+      select(cm, 5);
+
+      // when
+      editor.setVariables([ 'foobar', 'baz' ]);
+      startCompletion(cm);
+
+      // then
+      await expectEventually(() => {
+        const completions = currentCompletions(cm.state);
+        expect(completions).to.have.length(2);
+        expect(completions[0].label).to.have.eql('baz');
+      });
+    });
+
+
+    it('should change suggestion when variables are updated', async function() {
+
+      // given
+      const value = '{ "foo": "bar" }';
+
+      const editor = new JSONEditor();
+
+      editor.setValue(value);
+      editor.setVariables([ 'foobar', 'baz' ]);
+
+      const cm = getCm(editor);
+
+      // move cursor to the end
+      select(cm, 5);
+
+      // assume
+      startCompletion(cm);
+      await expectEventually(() => {
+        const completions = currentCompletions(cm.state);
+        expect(completions).to.have.length(1);
+        expect(completions[0].label).to.eql('foobar');
+      });
+
+      // when
+      editor.setVariables([ 'foobaz' ]);
+      startCompletion(cm);
+
+      // then
+      await expectEventually(() => {
+        const completions = currentCompletions(cm.state);
+        expect(completions).to.have.length(1);
+        expect(completions[0].label).to.eql('foobaz');
+      });
+    });
+  });
+
+
+  describe('autocompletion', function() {
+
+    it('should suggest applicable variables', function(done) {
+
+      // given
+      const initalValue = 'fooba';
+      const variables = [ 'foobar', 'baz' ];
+
+      const editor = new JSONEditor();
+      editor.setValue(initalValue),
+      editor.setVariables(variables);
+
+      const cm = getCm(editor);
+
+      // move cursor to the end
+      select(cm, 5);
+
+      // when
+      startCompletion(cm);
+
+      // then
+      // update done async
+      expectEventually(() => {
+        const completions = currentCompletions(cm.state);
+        expect(completions).to.have.length(1);
+        expect(completions[0].label).to.have.eql('foobar');
+        done();
+      });
+
+    });
+
+  });
+
+});
+
+
+// helper //////////////////////
+
+function select(editor, anchor, head = anchor) {
+  const cm = getCm(editor);
+
+  cm.dispatch({
+    selection: {
+      anchor,
+      head
+    }
+  });
+}
+
+function getCm(editor) {
+  return editor._view || editor;
+}
+
+/**
+ * Copied over from @bpmn-io/feel-editor.
+ */
+async function expectEventually(fn) {
```

Comment:
We could consider using React Testing Library, they have utilities like this too

Reply:
Good catch. We use Preact Testing Library which seems not to offer something like this. I think we should not mix up just because of some utilities. Also does not feel pretty natural to use React Testing utilities in a Preact library.

Reply:
Yeah, what I meant by React Testing Library was using the Preact counterpart.
It seems like the async hooks are actually part of the the `dom-testing-library` package, so it should be ok to use in Preact project https://testing-library.com/docs/dom-testing-library/api-async and they have examples in the Preact docs https://preactjs.com/guide/v10/preact-testing-library

Reply:
Thanks for the follow-up 👍 I'll see whether I can contribute a slight adjustment here on top.